core/classes/Placeholders.php

```diff
@@ -0,0 +1,162 @@
+<?php
+/*
+ *	Made by Aberdeener
+ *
+ *  https://github.com/NamelessMC/Nameless/
+ *  NamelessMC version 2.0.0-pr10
+ *
+ *  License: MIT
+ *
+ *  Plceholders class
+ */
+
+class Placeholders {
+
+    /** @var Placeholders */
+    private static $_instance = null;
+
+    /** @var DB */
+    private $_db = null;
+
+    private array $_all_placeholders;
+    private array $_leaderboard_data;
+
+    public static function getInstance() {
+        if(!isset(self::$_instance)) {
+            self::$_instance = new Placeholders();
+        }
+
+        return self::$_instance;
+    }
+
+    public function __construct() {
+        $this->_db = DB::getInstance();
+
+        $placeholders_query = $this->_db->get('placeholders_settings', ['name', '<>', ''])->results();
+        $placeholders = [];
+
+        foreach ($placeholders_query as $placeholder) {
+            $data = new stdClass();
+
+            $data->name = $placeholder->name;
+            $data->safe_name = sha1($placeholder->name);
+            $data->friendly_name = isset($placeholder->friendly_name) ? $placeholder->friendly_name : $placeholder->name;
+            $data->show_on_profile = $placeholder->show_on_profile;
+            $data->show_on_forum = $placeholder->show_on_forum;
+            $data->leaderboard = $placeholder->leaderboard;
+            $data->leaderboard_title = isset($placeholder->leaderboard_title) ? $placeholder->leaderboard_title : $data->friendly_name;
+            $data->leaderboard_sort = $placeholder->leaderboard_sort;
+            $data->leaderboard_settings_url = URL::build('/panel/core/placeholders', 'leaderboard=' . $data->safe_name);
+
+            $placeholders[] = $data;
+        }
+
+        $this->_all_placeholders = $placeholders;
+    }
+
+    /**
+     * Get all registered placeholders.
+     * 
+     * @return array All placeholders.
+     */
+    public function getAllPlaceholders() {
+        return $this->_all_placeholders;
+    }
+
+    /**
+     * Get placeholder data by name of placeholder.
+     * 
+     * @param string $placeholder_name Name of placeholder - must be hashed with sha1.
+     * @return object|null This placeholder's data, null if not exist.
+     */
+    public function getPlaceholderByName($placeholder_name) {
+        foreach ($this->_all_placeholders as $placeholder) {
+            if ($placeholder->safe_name == $placeholder_name) {
+                return $placeholder;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Create a new row in nl2_placeholders_settings if a row with the "name" of $name does not exist.
+     * 
+     * @param string $name Name of placeholder
+     */
+    public function registerPlaceholder($name) {
+        $this->_db->query("INSERT IGNORE INTO nl2_placeholders_settings (name) VALUES (?)", [$name]);
```

Comment:
Should we be using something other than `INSERT IGNORE` here? Updating an existing record instead maybe?

Reply:
Negatory, nl2_placeholders_settings contains 1 row per unique placeholder. All the registerPlaceholder function does is insert a row into that table, it does not update or insert any other information.